### PR TITLE
Sampling Oracles

### PR DESCRIPTION
**Pull request by vitorlppf**
An implementation of Sampling Oracles as suggested in #80, as well as renaming the `:noise` key to `:noiseidx` as suggested in #84.

I apologize for the delay. I wrote down most of the implementation a while ago, but I decided to wait some time before testing to do it with a clear mind. I ran into some minor issues, but I believe the current version should work. I tested it with [simplified_hydrothermal_dispatch.jl](https://github.com/odow/SDDP.jl/blob/master/examples/HydroValleys/simplified_hydrothermal_dispatch.jl).

### CODE REVIEW

**Comment by odow** [comment]
Awesome. Delays are fine! You are a volunteer. I'll have a better look later, but could you look into overloading `samplesubproblem` as well?

**Comment by vitorlppf** [comment]
I can certainly implement the current `samplesubproblem` method as the Default option for a subproblem sampling oracle. I think it would be better if those were of a different type from `SamplingOracle`. Do you agree? If so, I'm going to rename all the types added in the current implementation, for the sake of clarity.

File: src/SDDP.jl [hunk 1]
```diff
@@ -219,7 +219,8 @@ function forwardpass!(m::SDDPModel, settings::Settings, solutionstore=nothing)
 
         # choose and set RHS noise
         if hasnoises(sp)
-            (noiseidx, noise) = samplenoise(sp, solutionstore)
+            noise = samplenoise(settings.oracle, sp, solutionstore)
+            noiseidx = 0
             setnoise!(sp, noise)
```
Comment (odow) [inline src/SDDP.jl]: I think `samplenoise` should still return the tuple `(noiseidx, noise)`. Users should return the index if it exists in the model, or the index `0` if it is out-of-sample.

File: src/noises.jl [hunk 2]
```diff
@@ -13,20 +13,35 @@ function setnoise!(sp::JuMP.Model, noise::Noise)
...
-    noiseidx = sample(ext(sp).noiseprobability)
-    return noiseidx, ext(sp).noises[noiseidx]
-end
-samplenoise(sp::JuMP.Model, solutionstore::Void) = samplenoise(sp)
-function samplenoise(sp::JuMP.Model, solutionstore::Dict{Symbol, Any})
-    if length(solutionstore[:noise])>=ext(sp).stage
-        idx = solutionstore[:noise][ext(sp).stage]
+function samplenoise(oracle::SamplingOracle, sp::JuMP.Model, solutionstore::Dict{Symbol, Any})
+    if length(solutionstore[:noiseidx])>=ext(sp).stage
+        idx = solutionstore[:noiseidx][ext(sp).stage]
```
Comment (odow) [inline src/noises.jl]: This should probably go since the historical simulations can just use a DeterministicOracle.

File: src/simulate.jl [hunk 3]
```diff
@@ -77,14 +77,15 @@ markov state to sample in each stage.
 function simulate(m::SDDPModel,
```
Comment (odow) [inline src/simulate.jl]: This method should probably be replaced by one that uses a DeterministicOracle.

File: src/typedefinitions.jl [hunk 4]
```diff
@@ -231,6 +233,37 @@ Defaults to `false`.
 """
 MonteCarloSimulation(;frequency=0,min=20,max=0,step=1,confidence=0.95,termination=false) = MonteCarloSimulation(frequency,collect(min:step:max),confidence,termination)
 
+struct DefaultSamplingOracle <: SamplingOracle end
+
+struct UniformSamplingOracle <: SamplingOracle end
+
+struct DeterministicSamplingOracle{T} <: SamplingOracle 
+    data::Vector{T}
+end
```
Comment (odow) [inline src/typedefinitions.jl]: Can you add doc strings for these oracles like you have for CustomSamplingOracle?

File: src/utilities.jl [hunk 5]
```diff
@@ -156,7 +156,7 @@ end
 samplesubproblem(stage::Stage, last_markov_state::Int, solutionstore::Void) = samplesubproblem(stage, last_markov_state)
 
 function samplesubproblem(stage::Stage, last_markov_state, solutionstore::Dict{Symbol, Any})
-    if length(solutionstore[:noise]) >= stage.t
+    if length(solutionstore[:noiseidx]) >= stage.t
```
Comment (odow) [inline src/utilities.jl]: See my comment regarding using a DeterministicOracle for historical simulations.

File: src/noises.jl (hunk 2 above)
Comment (vitorlppf) [reply src/noises.jl]: Could you be more clear? Currently I intend to keep the existing methods under `DefaultSamplingOracle`, while having a separate method for `DeterministicSamplingOracle`.

File: src/utilities.jl (hunk 5 above)
Comment (vitorlppf) [reply src/utilities.jl]: When do you expect `length(solutionstore[:noiseidx])>=stage.t` to return `true` and when do you expect it to return `false`? It seems to me that the number of indexes stored would always be strictly lesser than the current stage during simulation, as I'd start at stage `1` with a 0-length index array and both would increment simultaneously inside [`forwardpass!`](https://github.com/odow/SDDP.jl/blob/master/src/SDDP.jl#L213).

File: src/noises.jl (hunk 2 above)
Comment (odow) [reply src/noises.jl]: The `solutionstore` was just a hack for running historical simulations. Now we could go
```julia
# Monte-Carlo
simulate(
    m, 
    100,
    oracle = DefaultOracle()
)

# Historical
simulate(
    m, 
    oracle = DeterministicOracle(
        markov_state = [1,1,2],
        noise_index = [1, 2, 1]
    )
)
```

So this whole method can probably be deleted.

File: src/utilities.jl (hunk 5 above)
Comment (odow) [reply src/utilities.jl]: This relates to https://github.com/odow/SDDP.jl/pull/88/files#r170729677 

Once `samplesubproblem` is implemented for the oralces, this method can be removed.

**Comment by odow** [comment]
p.s. can I ask where you are based any how you found out about the library?

**Pull request closed without merging**